### PR TITLE
script_pokemon_util_80A0058, battle_interface

### PR DESCRIPTION
**Pull request by PikalaxALT**

### CODE REVIEW

File: src/script_pokemon_util_80A0058.c [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        ppBonuses = GetMonData(&gPlayerParty[i], MON_DATA_PP_BONUSES);
+
+        // restore PP.
+        for(j = 0; j < MAX_MON_MOVES; j++)
+        {
+            arg[0] = CalculatePPWithBonus(GetMonData(&gPlayerParty[i], MON_DATA_MOVE1 + j), ppBonuses, j);
+            SetMonData(&gPlayerParty[i], MON_DATA_PP1 + j, arg);
+        }
+
+        // since status is u32, the four 0 assignments here are probably for safety to prevent undefined data from reaching SetMonData.
```
Comment (jiangzhengwenjz) [inline src/script_pokemon_util_80A0058.c]: What does the comment mean?

File: src/script_pokemon_util_80A0058.c [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    }
+
+    Free(mon);
+    return sentToPc;
+}
+
+u8 ScriptGiveEgg(u16 species)
+{
+    struct Pokemon *mon = AllocZeroed(sizeof(struct Pokemon));
+    u8 isEgg;
```
Comment (jiangzhengwenjz) [inline src/script_pokemon_util_80A0058.c]: `bool8`

File: src/script_pokemon_util_80A0058.c [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    int i;
+
+    // copy the selected pokemon according to the order.
+    for (i = 0; i < 3; i++)
+        if (gSelectedOrderFromParty[i]) // as long as the order keeps going (did the player select 1 mon? 2? 3?), do not stop
+            party[i] = gPlayerParty[gSelectedOrderFromParty[i] - 1]; // index is 0 based, not literal
+
+    CpuFill32(0, gPlayerParty, sizeof gPlayerParty);
+
+    // overwrite the first 4 with the order copied to.
```
Comment (jiangzhengwenjz) [inline src/script_pokemon_util_80A0058.c]: Isn't it 3?

File: src/battle_interface.c [hunk 4]
```diff
@@ -0,0 +1,2803 @@
...
+    s32 oldValue;
+    s32 receivedValue;
+    u32 pal:5;
+    u32 tileOffset;
+};
+
+enum
+{   // Corresponds to gHealthboxElementsGfxTable (and the tables after it) in graphics.c
+    // These are indexes into the tables, which are filled with 8x8 square pixel data.
+        HEALTHBOX_GFX_0, //hp bar [black section]
```
Comment (jiangzhengwenjz) [inline src/battle_interface.c]: inconsistent indent level

File: src/battle_interface.c [hunk 5]
```diff
@@ -0,0 +1,2803 @@
...
+{
+    s16 data6 = 0;
+    u8 healthboxLeftSpriteId;
+    u8 healthboxRightSpriteId;
+    u8 healthbarSpriteId;
+    struct Sprite *sprite;
+
+    if (!IsDoubleBattle())
+    {
+        if (GetBattlerSide(a) == 0)
```
Comment (jiangzhengwenjz) [inline src/battle_interface.c]: `B_SIDE_PLAYER`

File: src/battle_interface.c [hunk 6]
```diff
@@ -0,0 +1,2803 @@
...
+            data6 = 2;
+        }
+
+        gSprites[healthboxLeftSpriteId].oam.affineParam = healthboxRightSpriteId;
+        gSprites[healthboxRightSpriteId].hBar_HealthBoxSpriteId = healthboxLeftSpriteId;
+        gSprites[healthboxRightSpriteId].callback = SpriteCB_HealthBoxOther;
+    }
+    else
+    {
+        if (GetBattlerSide(a) == 0)
```
Comment (jiangzhengwenjz) [inline src/battle_interface.c]: `B_SIDE_PLAYER`

File: src/battle_interface.c [hunk 7]
```diff
@@ -0,0 +1,2803 @@
...
+            gSprites[healthboxRightSpriteId].hBar_HealthBoxSpriteId = healthboxLeftSpriteId;
+            gSprites[healthboxRightSpriteId].oam.tileNum += 32;
+            gSprites[healthboxRightSpriteId].callback = SpriteCB_HealthBoxOther;
+            data6 = 2;
+        }
+    }
+    healthbarSpriteId = CreateSpriteAtEnd(&gUnknown_82602F8[gBattlerPositions[a]], 140, 60, 0);
+    sprite = &gSprites[healthbarSpriteId];
+    SetSubspriteTables(sprite, &gUnknown_82603C4[GetBattlerSide(a)]);
+    sprite->subspriteMode = 2;
```
Comment (jiangzhengwenjz) [inline src/battle_interface.c]: `SUBSPRITES_IGNORE_PRIORITY`

File: src/battle_interface.c [hunk 8]
```diff
@@ -0,0 +1,2803 @@
...
+
+    if (!IsDoubleBattle())
+    {
+        if (GetBattlerSide(a) == 0)
+        {
+            healthboxLeftSpriteId = CreateSprite(&sHealthboxPlayerSpriteTemplates[0], 240, 160, 1);
+            healthboxRightSpriteId = CreateSpriteAtEnd(&sHealthboxPlayerSpriteTemplates[0], 240, 160, 1);
+
+            gSprites[healthboxLeftSpriteId].oam.shape = 0;
+            gSprites[healthboxRightSpriteId].oam.shape = 0;
```
Comment (jiangzhengwenjz) [inline src/battle_interface.c]: use `SPRITE_SHAPE`macro

File: src/battle_interface.c [hunk 9]
```diff
@@ -0,0 +1,2803 @@
...
+    return healthboxLeftSpriteId;
+}
+
+u8 CreateSafariPlayerHealthboxSprites(void)
+{
+    u8 healthboxLeftSpriteId = CreateSprite(&sHealthboxSafariSpriteTemplate, 240, 160, 1);
+    u8 healthboxRightSpriteId = CreateSpriteAtEnd(&sHealthboxSafariSpriteTemplate, 240, 160, 1);
+
+    gSprites[healthboxLeftSpriteId].oam.shape = ST_OAM_SQUARE;
+    gSprites[healthboxRightSpriteId].oam.shape = ST_OAM_SQUARE;
```
Comment (jiangzhengwenjz) [inline src/battle_interface.c]: use` SPRITE_SHAPE`macro

File: src/battle_interface.c [hunk 10]
```diff
@@ -0,0 +1,2803 @@
...
+}
+
+static const u8 gUnknown_826052C[20] = __("{COLOR 01}{HIGHLIGHT 02}");
+
+void UpdateHpTextInHealthbox(u8 healthboxSpriteId, s16 value, u8 maxOrCurrent)
+{
+    u32 windowId, spriteTileNum;
+    u8 *windowTileData;
+    u8 *strptr;
+    register void *objVram;
```
Comment (jiangzhengwenjz) [inline src/battle_interface.c]: required for matching?

File: src/battle_interface.c [hunk 11]
```diff
@@ -0,0 +1,2803 @@
...
+            TextIntoHealthboxObject((void *)(OBJ_VRAM0) + spriteTileNum * TILE_SIZE_4BPP + 0x2E0, windowTileData, 1);
+            TextIntoHealthboxObject((void *)(OBJ_VRAM0) + spriteTileNum * TILE_SIZE_4BPP + 0xA00, windowTileData + 0x20, 2);
+            RemoveWindowOnHealthbox(windowId);
+        }
+    }
+    else
+    {
+        u8 battler;
+
+        u8 text[20]; memcpy(text, gUnknown_826052C, sizeof(gUnknown_826052C));
```
Comment (jiangzhengwenjz) [inline src/battle_interface.c]: intentionally wrote on the same line?

File: src/script_pokemon_util_80A0058.c (hunk 1 above)
Comment (PikalaxALT) [reply src/script_pokemon_util_80A0058.c]: speculation as to why `arg` is set this way

File: src/script_pokemon_util_80A0058.c (hunk 3 above)
Comment (PikalaxALT) [reply src/script_pokemon_util_80A0058.c]: yes. i fixed the loops but not the comment lol

File: src/script_pokemon_util_80A0058.c [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+    heldItem[0] = item;
+    heldItem[1] = item >> 8;
+    SetMonData(mon, MON_DATA_HELD_ITEM, heldItem);
+    sentToPc = GiveMonToPlayer(mon);
+    nationalDexNum = SpeciesToNationalPokedexNum(species);
+
+    switch(sentToPc)
+    {
+    case 0:
+    case 1:
```
Comment (jiangzhengwenjz) [inline src/script_pokemon_util_80A0058.c]: are they `MON_GIVEN_TO_PARTY` and `MON_GIVEN_TO_PC`?

File: include/script_pokemon_util.h [hunk 13]
```diff
@@ -4,7 +4,7 @@
 bool8 ScriptGiveMon(u16 species, u8 level, u16 item, u32 unk1, u32 unk2, u8 unk3);
 bool8 ScriptGiveEgg(u16 species);
 void ScriptSetMonMoveSlot(u8 partyIdx, u16 move, u8 slot);
-void sp000_heal_pokemon(void);
+void HealPlayerParty(void);
 void ReducePlayerPartyToThree(void);
 void CreateScriptedWildMon(u16 species, u8 level, u16 item);
```
Comment (jiangzhengwenjz) [inline include/script_pokemon_util.h]: `GUARD_SCRIPT_POKEMON_UTIL_80A0058_H` -> `GUARD_SCRIPT_POKEMON_UTIL_H`

**Pull request merged**